### PR TITLE
Fixes #2916 - Actor Implementation of TaskTracker, TaskCreator, TaskUpdater

### LINKED ISSUE #2916: Actor-based implementation of the TaskTracker
Originally part of #462.

Acceptance Criteria 
- [ ] task state is held in an actor
- [ ] the actor uses the WhenLeader infrastructure
- [ ] reads should be served from memory, writes propagated to Zookeeper and only confirmed when Zookeeper confirmed the write 
- [ ] acceptable performance

### PR DESCRIPTION
**Pull request by kolloch**
Done (I hope ;) ).

* Also moved the TaskTracker, TaskUpdater, TaskCreator traits
  to the core packages (towards the new implementation)
* Got rid of the recently introduced TaskReconciler because
  it is not needed anymore with the new implementation. Supposedly
  for concurrency reasons, the old implementation had a concept
  "Apps" in the TaskTracker. In the new implementation,
  tasks are grouped by app id but app ids are not retained when no
  tasks are known for the app anymore.
* Introduced async variants for all TaskTracker methods to encourage
  usage of asynchronous programming. I would like it better
  to rename the synchronous methods to "*Sync" but that would
  have conflated the size of this PR.

### CODE REVIEW

**Comment by kolloch** [comment]
rebased and added a TaskTrackerDelegate metric

**Comment by kolloch** [comment]
The test error is probably related to https://github.com/mesosphere/marathon/pull/2931

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerConfig.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package mesosphere.marathon.core.task.tracker
+
+import org.rogach.scallop.ScallopConf
+
+trait TaskTrackerConfig extends ScallopConf {
+  //scalastyle:off magic.number
+
+  lazy val taskTrackerRequestTimeout = opt[Int]("task_tracker_request_timeout",
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerConfig.scala]: can we rename this parameter to internalXXX so that we easily see it is internal and hidden by default?

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerConfig.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    descr = "INTERNAL TUNING PARAMETER: Timeout (in ms) for requests to the taskTracker.",
+    hidden = true,
+    default = Some(10000))
+
+  lazy val taskUpdateRequestTimeout = opt[Int]("task_update_request_timeout",
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerConfig.scala]: same as above: internalXXX

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  *
+  * FIXME: To allow introducing the new asynchronous [[TaskTracker]] without needing to
+  * refactor a lot of code at once, synchronous methods are still available.
+  */
+trait TaskTracker {
+
+  def getTasks(appId: PathId): Iterable[MarathonTask]
+  def getTasksAsync(appId: PathId)(implicit ec: ExecutionContext): Future[Iterable[MarathonTask]]
+
+  def getTask(appId: PathId, taskId: String): Option[MarathonTask]
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala]: We probably should add @deprecated here?

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+class TaskTrackerModule(
+    clock: Clock,
+    metrics: Metrics,
+    config: TaskTrackerConfig,
+    leadershipModule: LeadershipModule,
+    taskRepository: TaskRepository) {
+  lazy val taskTracker: TaskTracker = new TaskTrackerDelegate(Some(metrics), config, taskTrackerActorRef)
+
+  def taskUpdater: TaskUpdater = taskTrackerCreatorAndUpdater
+  def taskCreator: TaskCreator = taskTrackerCreatorAndUpdater
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala]: I think I can not create tasks with the TaskCreator - what about  `TaskCreationListener` or equivalent to better match the nature?

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoader.scala [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package mesosphere.marathon.core.task.tracker.impl
+import scala.concurrent.Future
+
+/**
+  * Loads all task data into an [[AppDataMap]].
+  */
+private[tracker] trait TaskLoader {
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoader.scala]: Generally I like traits very much. Do we want to ouse them also internal object that are not used out of this context?

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  */
+private[tracker] class TaskLoaderImpl(repo: TaskRepository) extends TaskLoader {
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  private[this] val log = LoggerFactory.getLogger(getClass.getName)
+
+  override def loadTasks(): Future[AppDataMap] = {
+    for {
+      names <- repo.allIds()
+      _ = log.info(s"About to load ${names.size} tasks")
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala]: debug?

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      tasks <- Future.sequence(names.map(repo.task(_))).map(_.flatten)
+    } yield {
+      log.info(s"Loaded ${tasks.size} tasks")
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala]: debug?

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+  override def process(op: Operation)(implicit ec: ExecutionContext): Future[Unit] = {
+    op.action match {
+      case Action.Update(task) => // used for a create or as a result from a UpdateStatus action
+        repo.store(task).map { _ =>
+          taskTrackerRef ! TaskTrackerActor.TaskUpdated(op.appId, task, TaskTrackerActor.Ack(op.sender))
+        }.recoverWith(tryToRecover(op)(expectedTaskState = Some(task)))
+      case Action.Expunge => // used for task termination or as a result from a UpdateStatus action
+        repo.expunge(op.taskId).map { _ =>
+          taskTrackerRef ! TaskTrackerActor.TaskRemoved(op.appId, op.taskId, TaskTrackerActor.Ack(op.sender))
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: The action has a sender. The sender is informed for Noop/Fail but not for update /Expunge/UpdateStatus.
Why?
Perhaps add this as comment?

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  */
+private[tracker] class TaskOpProcessorImpl(
+    taskTrackerRef: ActorRef,
+    repo: TaskRepository,
+    statusUpdateActionResolver: StatusUpdateActionResolver) extends TaskOpProcessor {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  import TaskOpProcessor._
+
+  override def process(op: Operation)(implicit ec: ExecutionContext): Future[Unit] = {
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: Should we check the deadline of this operation?

File: src/main/scala/mesosphere/marathon/state/Migration.scala [hunk 10]
```diff
@@ -42,14 +37,10 @@ class Migration @Inject() (
     */
   def migrations: List[MigrationAction] = List(
     StorageVersions(0, 5, 0) -> { () =>
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/state/Migration.scala]: Guess jou can remove it for 0.5, if 0.7 is not supported.

File: src/main/scala/mesosphere/marathon/state/Migration.scala [hunk 11]
```diff
@@ -42,14 +37,10 @@ class Migration @Inject() (
...
-      changeApps(app => app.copy(id = app.id.toString.toLowerCase.replaceAll("_", "-").toRootPath))
+      Future.failed(new IllegalStateException("migration from 0.5.x not supported anymore"))
     },
     StorageVersions(0, 7, 0) -> { () =>
-      for {
-        _ <- changeTasks(app => new InternalApp(app.appName.canonicalPath(), app.tasks, app.shutdown))
-        _ <- changeApps(app => app.copy(id = app.id.canonicalPath()))
-        _ <- putAppsIntoGroup()
-      } yield ()
+      Future.failed(new IllegalStateException("migration from 0.7.x not supported anymore"))
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/state/Migration.scala]: Yay.

**Comment by aquamatthias** [comment]
Looks good. 
I left a couple of comments.
Please also rebase.

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerConfig.scala (hunk 1 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerConfig.scala]: nice suggestion

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala (hunk 3 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala]: Then the caller code will not compile anymore with our settings.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoader.scala (hunk 5 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoader.scala]: What would you suggest instead? Just mock the Impl directly? I like this approach better -- not much more work -- but I do not care that much. If you want to, I can change this.

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala (hunk 4 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala]: I don't like the name either. But, unfortunately, I cannot think of something better. `TaskCreationListener` sounds to passive. The methods in `TaskCreator` do result in the task being "created/stored" in the `TaskTracker` or a task being removed from the TaskTracker.

Better suggestions?

Maybe we should keep the renaming to another PR, though.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala (hunk 6 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala]: It is displayed once on startup. I would keep it at info. You do see if Marathon was stuck loading the tasks that way.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala (hunk 7 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala]: See above.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 8 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: If we have to update the state of the TaskTracker, we have to go through the TaskTracker for acknowledging the change (See Ack). If we did not change anything, then we can inform the sender directly.

I'll try to come up with good documentation for that.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 9 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: This method is only called by the `TaskUpdateActor`. I chose to do the deadline checking there but the decision was quite arbitrary -- I just avoid a context switch there.

File: src/main/scala/mesosphere/marathon/state/Migration.scala (hunk 10 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/Migration.scala]: I think that you are right.

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala (hunk 4 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala]: `taskCreationHandler`?

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala (hunk 3 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala]: True - probably too much hassle in the moment.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoader.scala (hunk 5 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoader.scala]: Not important: since this is a simple helper class, I wonder if we need a trait for that.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala (hunk 6 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskLoaderImpl.scala]: OK

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 8 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: Understant. Comment would help.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 9 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: I would expect that check in process(Operation).  Perhaps add a line of documentation?

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala (hunk 4 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTrackerModule.scala]: ack

**Comment by kolloch** [comment]
Thanks for the review @aquamatthias. I addressed your comments in four additional commits and also rebased the PR to the latest master.

**Comment by aquamatthias** [comment]
@kolloch Failing test.

**Comment by kolloch** [comment]
Sorry, forgot to commit/push my test fix.

**Comment by aquamatthias** [comment]
Thanks @kolloch can you squash the commits?
